core/src/main/scala/org/apache/spark/rdd/SortMergeJoinRDD.scala

```diff
@@ -0,0 +1,281 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.rdd
+
+import scala.reflect.ClassTag
+
+import org.apache.spark._
+import org.apache.spark.serializer.Serializer
+import org.apache.spark.shuffle.ShuffleHandle
+import org.apache.spark.util.collection.CompactBuffer
+
+private[spark] case class ShuffleJoinSplitDep(handle: ShuffleHandle) extends Serializable
+
+private[spark] class JoinPartition(
+    idx: Int,
+    val left: ShuffleJoinSplitDep,
+    val right: ShuffleJoinSplitDep)
+  extends Partition with Serializable {
+  override val index: Int = idx
+  override def hashCode(): Int = idx
+}
+
+private[spark] abstract class JoinType[K: ClassTag,
+                                          L: ClassTag,
+                                          R: ClassTag,
+                                          PAIR <: Product2[_, _]]
+    extends Serializable {
+
+  private[spark] var joinType: Int = 0
+
+  def flatten(iterators: Iterator[(K, (Iterator[L], Iterator[R]))]): Iterator[(K, PAIR)]
+
+  def mergeIterators(leftIter: Iterator[L], rightIter: Iterator[R]): Iterator[(L, R)] = {
```

Comment:
Not saying that it is wrong, just curious, is it any different from ziping iterators?